File: server/php/src/Infrastructure/Mappers/MandateMapper.php
<?php

namespace OnlinePayments\ExampleApp\Infrastructure\Mappers;

use OnlinePayments\ExampleApp\Application\Domain\Enums\Common\Country;
use OnlinePayments\ExampleApp\Application\Domain\Payments\Mandate;
use OnlinePayments\ExampleApp\Application\DTOs\Mandate\GetMandateResponseDto;
use OnlinePayments\ExampleApp\Application\DTOs\Payment\RequestDto;
use OnlinePayments\ExampleApp\Application\Extensions\Models\CountryExtension;
use OnlinePayments\Sdk\Domain\BankAccountIban;
use OnlinePayments\Sdk\Domain\CreateMandateRequest;
use OnlinePayments\Sdk\Domain\CreateMandateResponse;
use OnlinePayments\Sdk\Domain\GetMandateResponse;
use OnlinePayments\Sdk\Domain\MandateAddress;
use OnlinePayments\Sdk\Domain\MandateCustomer;
use OnlinePayments\Sdk\Domain\MandatePersonalInformation;
use OnlinePayments\Sdk\Domain\MandatePersonalName;

final class MandateMapper
{
    public static function mapToCreateMandateRequest(?RequestDto $requestDto): CreateMandateRequest
    {
        $mandate = $requestDto?->mandate;

        $createMandate = new CreateMandateRequest();

        $customer = new MandateCustomer();

        $bankAccount = new BankAccountIban();
        $bankAccount->iban = $mandate?->iban;
        $customer->bankAccountIban = $bankAccount;

        $address = new MandateAddress();
        $address->countryCode = $mandate?->address?->country !== null
            ? CountryExtension::toIsoAlpha2(Country::tryFrom($mandate->address->country))
            : null;
        $address->city = $mandate?->address?->city;
        $address->street = $mandate?->address?->street;
        $address->zip = $mandate?->address?->zip;
        $customer->mandateAddress = $address;

        $personalInformation = new MandatePersonalInformation();
        $name = new MandatePersonalName();
        $name->firstName = $mandate?->address?->firstName;
        $name->surname = $mandate?->address?->lastName;
        $personalInformation->name = $name;
        $customer->personalInformation = $personalInformation;

        $createMandate->customer = $customer;
        $createMandate->customerReference = $mandate?->customerReference;
        $createMandate->recurrenceType = $mandate?->recurrenceType;
        $createMandate->returnUrl = $mandate?->returnUrl;
        $createMandate->signatureType = $mandate?->signatureType;

        return $createMandate;
    }

    public static function mapFromGetMandateResponse(?GetMandateResponse $getMandateResponse): ?GetMandateResponseDto
    {
        if ($getMandateResponse?->mandate === null) {
            return null;
        }

        return new GetMandateResponseDto($getMandateResponse->mandate->uniqueMandateReference);
    }

    public static function mapFromCreateMandateResponse(CreateMandateResponse $createMandateResponse): Mandate
    {
        $mandate = new Mandate();

        $mandate->iban = $createMandateResponse->mandate->customer->bankAccountIban->iban;
        $mandate->customerReference = $createMandateResponse->mandate->customerReference;
        $mandate->returnUrl = $createMandateResponse->merchantAction->redirectData->redirectURL;

        return $mandate;
    }
}

File: server/ruby/app/infrastructure/mappers/mandate_mapper.rb
module Infrastructure
  module Mappers
    module MandateMapper
      module_function

      def to_sdk_request(request_dto)
        mandate = request_dto&.mandate

        create_mandate = OnlinePayments::SDK::Domain::CreateMandateRequest.new

        customer = OnlinePayments::SDK::Domain::MandateCustomer.new

        bank_account_iban = OnlinePayments::SDK::Domain::BankAccountIban.new
        bank_account_iban.iban = mandate&.iban
        customer.bank_account_iban = bank_account_iban

        address = OnlinePayments::SDK::Domain::MandateAddress.new
        if mandate&.address&.country
          address.country_code = Business::Extensions::Models::CountryExtension.to_iso_alpha2(mandate.address.country)
        end
        address.city = mandate&.address&.city
        address.street = mandate&.address&.street
        address.zip = mandate&.address&.zip
        customer.mandate_address = address

        personal_information = OnlinePayments::SDK::Domain::MandatePersonalInformation.new
        name = OnlinePayments::SDK::Domain::MandatePersonalName.new
        name.first_name = mandate&.address&.first_name
        name.surname = mandate&.address&.last_name
        personal_information.name = name
        customer.personal_information = personal_information

        create_mandate.customer = customer

        create_mandate.customer_reference = mandate&.customer_reference
        create_mandate.recurrence_type = mandate&.recurrence_type
        create_mandate.return_url = mandate&.return_url
        create_mandate.signature_type = mandate&.signature_type

        create_mandate
      end

      def from_sdk_get_response(get_mandate_response)
        return nil if get_mandate_response&.mandate.nil?

        Business::Dtos::Mandate::GetMandateResponseDto.new(
          get_mandate_response.mandate.unique_mandate_reference
        )
      end

      def from_sdk_create_response(create_mandate_response)
        mandate = Business::Domain::Payments::Mandate.new

        mandate.iban = create_mandate_response.mandate.customer.bank_account_iban.iban
        mandate.customer_reference = create_mandate_response.mandate.customer_reference
        mandate.return_url = create_mandate_response.merchant_action.redirect_data.redirect_url

        mandate
      end
    end
  end
end